dask/array/tests/test_array_core.py

```diff
@@ -1210,6 +1210,16 @@ def foo(x):
     assert 'RuntimeError' in msg
 
 
+def test_map_blocks_no_array_args():
+    def func(dtype, block_info=None):
+        loc = block_info[None]['array-location']
+        return np.arange(loc[0][0], loc[0][1], dtype=dtype)
+
+    x = da.map_blocks(func, np.float32, new_axis=[0], chunks=((5, 3),), dtype=np.float32)
+    assert x.chunks == ((5, 3),)
+    assert_eq(x, np.arange(8, dtype=np.float32))
```

Comment:
I tried using this naively and came up with.

```python
In [26]: def f():
    ...:     return np.ones((2, 3), np.float32)
    ...:

In [27]: x = da.map_blocks(f, chunks=((2, 2, 2), (3, 3)))
```

```python-traceback
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-27-f1c3322fbb11> in <module>
----> 1 x = da.map_blocks(f, chunks=((2, 2, 2), (3, 3)))

~/workspace/dask/dask/array/core.py in map_blocks(func, *args, **kwargs)
    523         if len(chunks) != len(out.numblocks):
    524             raise ValueError("Provided chunks have {0} dims, expected {1} "
--> 525                              "dims.".format(len(chunks), len(out.numblocks)))
    526         chunks2 = []
    527         for i, (c, nb) in enumerate(zip(chunks, out.numblocks)):

ValueError: Provided chunks have 2 dims, expected 0 dims.
```

I then realized that I needed to provide the `new_axis=` keyword argument (by looking at this test)

```python
x = da.map_blocks(f, chunks=((2, 2, 2), (3, 3)), new_axis=[0, 1])
```

But I can imagine that it'd be nicer not to have to do this.  No obligation to handle this, but do you have any thoughts on nice ways to improve this for users?

Reply:
That bugged me too. I wanted to come up with something generic rather than a special rule for the zero-input case.

How about the following: if `chunks` is specified and `new_axis` is not specified, then `new_axis` will default to insert the appropriate number of axes on the left (inserting new axes on the left is consistent with numpy's broadcasting rules for mixed-dimensionality operations).

If you're happy with that, I'll check the code to figure out what "appropriate number" actually means in code, and make sure it interacts sensibly with `drop_axis`.

Reply:
That sounds sensible to me (though I wouldn't be surprised if there were odd complications) I say lets go for it!  (assuming you're willing to do all the work :) )